File: java/src/lags/LagsService.java
package lags;

import java.io.*;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class LagsService {

    List<Ordre> listeOrdre = new ArrayList<>();

    // lit le fihier des ordres et calcule le CA
    public void getFichierOrder(String fileName)
    {
        try
        {
            File file = new File(fileName);
            FileReader fr = new FileReader(file);
            BufferedReader br = new BufferedReader(fr);
            String line;
            while((line = br.readLine()) != null){
                String[] champs = line.split(";");
                String chp1 = champs[0];
                int chp2 = Integer.parseInt(champs[1]);
                int champ3 = Integer.parseInt(champs[2]);
                double chp4 = Double.parseDouble(champs[3]);
                Ordre ordre = new Ordre(chp1, chp2, champ3, chp4);
                listeOrdre.add(ordre);
            }
            br.close();
            fr.close();
        }
        catch (FileNotFoundException e)
        {
            System.out.println("FICHIER ORDRES.CSV NON TROUVE. CREATION FICHIER.");
            writeOrdres(fileName);
        } catch (IOException e) {
            // TODO CRU à toi de voir ce que tu fais ici
        }
    }
    // écrit le fichier des ordres
    void writeOrdres(String nomFich)
    {
        try {
            FileWriter writer = new FileWriter(new File(nomFich));
            for (Ordre ordre : listeOrdre)
            {
                String[] ligneCSV = new String[4];
                ligneCSV[0] = ordre.getId();
                ligneCSV[1] = Integer.toString(ordre.getDebut());
                ligneCSV[2] = Integer.toString(ordre.getDuree());
                ligneCSV[3] = Double.toString(ordre.getPrix());
                writer.write(String.join(";", ligneCSV) + "\n");
            }
            writer.close();
        } catch (IOException e) {
            // TODO CRU à toi de voir ce qu'on fait ici
        }
    }

    // affiche la liste des ordres
    public void liste()
    {
        System.out.println("LISTE DES ORDRES");
        System.out.println(String.format("%-8s %10s %5s %10s","ID", "DEBUT", "DUREE", "PRIX"));
        System.out.println(String.format("%-8s %10s %5s %10s", "--------", "-------", "-----", "----------"));
        listeOrdre.stream().sorted((o1,o2) -> Integer.compare(o1.getDebut(), o2.getDebut())).forEach(this::afficherOrdre);
        System.out.println(String.format("%-8s %10s %5s %10s","--------", "-------", "-----", "----------"));
    }

    public void afficherOrdre(Ordre ordre)
    {
        System.out.println(String.format("%-8s %10d %5d %10f", ordre.getId(), ordre.getDebut(), ordre.getDuree(), ordre.getPrix()));
    }
    // Ajoute un ordre; le CA est recalculé en conséquence
    public void ajouterOrdre()
    {
        System.out.println("AJOUTER UN ORDRE");
        System.out.println("FORMAT = ID;DEBUT;FIN;PRIX");
        String line = System.console().readLine();
        line = line.toUpperCase();
        String[] champs = line.split(";");
        String id = champs[0];
        int dep = Integer.parseInt(champs[1]);
        int dur = Integer.parseInt(champs[2]);
        double prx = Double.parseDouble(champs[3]);
        Ordre ordre = new Ordre(id, dep, dur, prx);
        listeOrdre.add(ordre);
        writeOrdres("..\\ordres.csv");
    }

    //public void CalculerLeCA()
    //{
    //    System.out.println("CALCUL CA..");
    //    laListe = laListe.OrderBy(ordre => ordre.debut).ToList();
    //    double ca = CA(laListe);
    //    System.out.println("CA: {0,10:N2}", ca);
    //}

    private double CA(List<Ordre> ordres, boolean debug)
    {
        // si aucun ordre, job done, TROLOLOLO..
        if (ordres.size() == 0)
            return 0.0;
        Ordre order = ordres.get(0);
        // attention ne marche pas pour les ordres qui depassent la fin de l'année
        // voir ticket PLAF nO 4807
        List<Ordre> liste = ordres.stream().filter(o -> o.getDebut() >= (o.getDebut() + o.getDuree())).collect(Collectors.toList());
        List<Ordre> liste2 = ordres.subList(1, ordres.size() - 1);
        double ca = order.getPrix() + CA(liste, debug);
        // Lapin compris?
        double ca2 = CA(liste2, debug);
        System.out.println(debug ? String.format("{0,10:N2}\n", Math.max(ca, ca2)) : ".");
        return Math.max(ca, ca2); // LOL
    }

    // MAJ du fichier
    public void suppression()
    {
        System.out.println("SUPPRIMER UN ORDRE");
        System.out.println("ID:");
        String id = System.console().readLine();
        this.listeOrdre = listeOrdre.stream().filter(o -> !o.getId().equals(id.toUpperCase())).collect(Collectors.toList());
        writeOrdres("..\\ORDRES.CSV");
    }

    // internal en java ???
    /*internal*/ void calculerLeCA(boolean debug)
    {
        System.out.println("CALCUL CA..");
        listeOrdre = listeOrdre
                .stream()
                .sorted((o1,o2) -> Integer.compare(o1.getDebut(), o2.getDebut()))
                .collect(Collectors.toList());
        double ca = CA(listeOrdre, debug);
        System.out.printf("CA: %10d", ca);
    }

}

File: CSharp/Lags/LagsService.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;

namespace Lags
{
    class LagsService
    {
        List<Order> ListOrder = new List<Order>();

        // read the order fil and calculate GS
        public void getFichierOrder(String fileName)
        {
            try
            {
                using (var reader = new StreamReader(fileName))
                {
                    while (!reader.EndOfStream)
                    {
                        var fields = reader.ReadLine().Split(';');
                        String fld1 = fields[0];
                        int fld2 = Int32.Parse(fields[1]);
                        int field3 = Int32.Parse(fields[2]);
                        double fld4 = Double.Parse(fields[3]);
                        Order order = new Order(fld1, fld2, field3, fld4);
                        ListOrder.Add(order);
                    }
                }
            }
            catch (FileNotFoundException e)
            {
                Console.WriteLine("CSV ORDER FILE NOT FOUND: CREATING ONE.");
                WriteOrders(fileName);
            }
        }
        // write order file
        void WriteOrders(String fileNam)
        {
            using (TextWriter writer = File.CreateText(fileNam))
            {
                foreach (Order order in ListOrder)
                {
                    string[] ligneCSV = new string[4];
                    ligneCSV[0] = order.id;
                    ligneCSV[1] = order.Start.ToString();
                    ligneCSV[2] = order.Duration.ToString();
                    ligneCSV[3] = order.Price.ToString();
                    writer.WriteLine(string.Join(";", ligneCSV));
                }
            }
        }


        // show order list
        public void List()
        {
            Console.WriteLine("ORDERS LIST");
            Console.WriteLine("{0,-8} {1,7} {2,5} {3,10}",
                "ID", "START", "DURATION", "PRICE");
            Console.WriteLine("{0,-8} {1:0000000} {2:00000} {3,10:N2}",
               "--------", "-------", "-----", "----------");
            ListOrder = ListOrder.OrderBy(order => order.Start).ToList();
            ListOrder.ForEach(ShowOrder);
            Console.WriteLine("{0,-8} {1:0000000} {2:00000} {3,10:N2}",
               "--------", "-------", "-----", "----------");
        }

        public void ShowOrder(Order order)
        {
            Console.WriteLine("{0,-8} {1:0000000} {2:00000} {3,10:N2}",
                order.id, order.Start, order.Duration, order.Price);

        }
        // Add order; GS is recalculated
        public void AddOrder()
        {
            Console.WriteLine("ADD ORDER");
            Console.WriteLine("FORMAT = ID;START;END;PRICE");
            String line = Console.ReadLine().ToUpper();
            var fields = line.Split(';');
            String id = fields[0];
            int star = Int32.Parse(fields[1]);
            int dur = Int32.Parse(fields[2]);
            double pr = Double.Parse(fields[3]);
            Order order = new Order(id, star, dur, pr);
            ListOrder.Add(order);
            WriteOrders("ordres.csv");
        }

        //public void CalculerTheGS()
        //{
        //    Console.WriteLine("CALCUL.  GS..");
        //    laListe = laListe.OrderBy(ordre => ordre.debut).ToList();
        //    double ca = CA(laListe);
        //    Console.WriteLine("CA: {0,10:N2}", ca);
        //}

        private double GS(List<Order> orders, bool debug)
        {
            // no order, job done, TROLOLOLO..
            if (orders.Count() == 0)
                return 0.0;
            Order order = orders.ElementAt(0);
            // attention ne marche pas pour les ordres qui depassent la fin de l'année 
            // voir ticket PLAF nO 4807 
            List<Order> list = orders.Where(orderr => orderr.Start >= order.Start + order.Duration).ToList();
            List<Order> list2 = orders.GetRange(1, orders.Count() - 1);
            double gs = order.Price + GS(list, debug);
            // I can live... with or withooooooout youuuuuu
            double gs2 = GS(list2, debug);
            Console.Write(debug ? String.Format("{0,10:N2}\n", Math.Max(gs, gs2)):".");
            return Math.Max(gs, gs2); // LOL
        }

        // File Update
        public void Delete()
        {
            Console.WriteLine("DELETE ORDER");
            Console.Write("ID:");
            string id = Console.ReadLine().ToUpper();
            this.ListOrder = ListOrder.Where(ordre => ordre.id != id).ToList();
            WriteOrders("ORDRES.CSV");
        }



        internal void CalculateTheGS(bool debug)
        {
            Console.WriteLine("CALCUL CA..");
            ListOrder = ListOrder.OrderBy(ordre => ordre.Start).ToList();
            double ca = GS(ListOrder, debug);
            Console.WriteLine("CA: {0,10:N2}", ca);
        }

    }
}
